### PR TITLE
Added urbd trigger, closes #14

### LINKED ISSUE #14: Add urban dictionary support

### PR DESCRIPTION
**Pull request by m-242**
Signed-off-by: Olivier Moreau m242 <m242@protonmail.com>

### CODE REVIEW

**Comment by m-242** [comment]
Urban dictionnary support.

File: bitbot/urbandict.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	Downvote   int `json:"thumbs_down"`
+}
+
+func Query(searchTerm string) (*SearchResult, error) {
+	const API_URL = "http://api.urbandictionary.com/v0/define?term="
+	resp, err := http.Get(API_URL + url.QueryEscape(searchTerm))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
```
[the remark this answers is not in the document]
Comment (m-242) [reply bitbot/urbandict.go:73]: false positive

File: bitbot/urbandict.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/whyrusleeping/hellabot"
+)
+
+var UrbanDictionnaryTrigger = NamedTrigger{
```
[the remark this answers is not in the document]
Comment (bbriggs) [reply bitbot/urbandict.go]: Spelling

File: bitbot/urbandict.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	Word       string
+	Definition string
+	Example    string
+	Permalink  string
+	Upvote     int `json:"thumbs_up"`
+	Downvote   int `json:"thumbs_down"`
+}
+
+func Query(searchTerm string) (*SearchResult, error) {
+	const API_URL = "http://api.urbandictionary.com/v0/define?term="
```
[the remark this answers is not in the document]
Comment (bbriggs) [reply bitbot/urbandict.go]: Why set an all-caps const inside a function?

File: bitbot/urbandict.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	"strings"
+
+	"github.com/whyrusleeping/hellabot"
+)
+
+var UrbanDictionaryTrigger = NamedTrigger{
+	ID:   "urbandict",
+	Help: "Get an urban dictionary issued definition. Usage: !urbd [term]",
+	Condition: func(irc *hbot.Bot, m *hbot.Message) bool {
+		return m.Command == "PRIVMSG" && strings.HasPrefix(m.Trailing, "!urbd")
```
Comment (bbriggs) [inline bitbot/urbandict.go:20]: Let's make the trigger `!ud` instead

File: bitbot/urbandict.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+type SearchResult struct {
+	Type    string `json:"result_type"`
+	Tags    []string
+	Results []Result `json:"list"`
+	Sounds  []string
+}
+
+type Result struct {
```
[the remark this answers is not in the document]
Comment (bbriggs) [reply bitbot/urbandict.go:57]: This is a really generic struct name and we will have name collisions if we leave it this way. Let's make it something. Also, I don't think this needs to be exported.

File: bitbot/urbandict.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+func cleanDef(def string) string {
+	def = strings.ReplaceAll(def, "[", "")
+	def = strings.ReplaceAll(def, "]", "")
+
+	return def
+}
+
+type SearchResult struct {
```
[the remark this answers is not in the document]
Comment (bbriggs) [reply bitbot/urbandict.go:50]: Unless `encoding/json`  needs this exported, we shouldn't export this struct

File: bitbot/urbandict.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+func urbanDictQuery(searchTerm string) (*SearchResult, error) {
+	const baseURL = "http://api.urbandictionary.com/v0/define?term="
+	resp, err := http.Get(baseURL + url.QueryEscape(searchTerm))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
```
Comment (bbriggs) [inline bitbot/urbandict.go:75]: this reminds me that we should make a generic HTTP client with proper timeouts that can be used for all of our API lookups

**Pull request merged**